src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -411,6 +447,149 @@ void ReduceAxesCompute(const nnvm::NodeAttrs& attrs,
+  if (req == kNullOp) return;
+  const ReduceAxesParam& param = nnvm::get<ReduceAxesParam>(attrs.parsed);
+  CHECK_EQ(param.axis.ndim(), 1U) << "sum(csr) only supports axis 0 or 1";
+  CHECK(param.axis[0] == 0 || param.axis[0] == 1)
+      << "sum(csr) only support axis 0 or 1";
+  CHECK(!param.keepdims) << "keepdims not supported for sparse";
+  CHECK(!param.exclude) << "exclude not supported for sparse";
+  int64_t out_data_size = 0;
+  if (param.axis[0] == 0) {
+    out_data_size = input.shape()[1];
+  } else {
+    out_data_size = input.shape()[0];
+  }
+  // only dense output storage type is supported
+  CHECK_EQ(output->storage_type(), kDefaultStorage);
+
+  CHECK_NE(req, kWriteInplace);
+
+  using namespace mshadow;
+  using namespace mxnet_op;
+  using namespace csr;
+  if (req == kWriteTo) {
+    MSHADOW_TYPE_SWITCH(output->data().type_flag_, DType, {
+      Kernel<set_zero, xpu>::Launch(s, out_data_size,
+                                    output->data().dptr<DType>());
+    })
+  }
+
+  if (!input.storage_initialized()) {
+    return;
+  }
+
+  const int64_t num_rows = input.shape()[0];
+
+  if (0 == param.axis[0]) {
+    MSHADOW_IDX_TYPE_SWITCH(input.aux_type(kIndPtr), RType, {
+      MSHADOW_IDX_TYPE_SWITCH(input.aux_type(kIdx), IType, {
+        MSHADOW_TYPE_SWITCH(input.dtype(), DType, {
+          MXNET_ASSIGN_REQ_SWITCH(req, req_type, {
+            auto in_indptr = input.aux_data(kIndPtr).dptr<RType>();
+            auto in_idx = input.aux_data(kIdx).dptr<IType>();
+            auto in_data = input.data().dptr<DType>();
+            Kernel<SumCsrKernel<req_type, 0>, xpu>::Launch(
+                s, out_data_size, output->data().dptr<DType>(), in_indptr,
+                in_idx, in_data, num_rows);
+          });
+        });
+      });
+    });
+  } else if (1 == param.axis[0]) {
+    MSHADOW_IDX_TYPE_SWITCH(input.aux_type(kIndPtr), RType, {
+      MSHADOW_TYPE_SWITCH(input.dtype(), DType, {
+        MXNET_ASSIGN_REQ_SWITCH(req, req_type, {
+          auto in_indptr = input.aux_data(kIndPtr).dptr<RType>();
+          auto in_data = input.data().dptr<DType>();
+          Kernel<SumCsrKernel<req_type, 1>, xpu>::Launch(
+              s, out_data_size, output->data().dptr<DType>(), in_indptr,
+              in_data);
+        });
+      });
+    });
+  }
+}
+
+template <typename xpu, typename reducer, bool normalize = false>
+void SumOpForwardEx(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
+                    const std::vector<NDArray>& inputs,
+                    const std::vector<OpReqType>& req,
+                    const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  mshadow::Stream<xpu>* s = ctx.get_stream<xpu>();
+  const NDArrayStorageType istype = inputs[0].storage_type();
+  if (istype == kCSRStorage) {
+    CHECK_EQ(inputs[0].shape().ndim(), 2U) << "sum(csr) op only supports"
```

Comment:
You don’t need the other <<. In fact, this causes an extra function call every time this function is executed regardless of whether the check succeeds or fails, so try to keep number of << calls low.

Reply:
fixed.